lib/puppet/type/user.rb

```diff
@@ -459,40 +467,47 @@ def insync?(current)
       end
     end
 
-    # Autorequire the group, if it's around
+    # Autorequire groups, if they're around
     autorequire(:group) do
       autos = []
 
-      # autorequire primary group, if managed
-      obj = @parameters[:gid]
-      groups = obj.shouldorig if obj
-      if groups
-        groups = groups.collect { |group|
-          if group.is_a?(String) && group =~/^\d+$/
-            Integer(group)
-          else
-            group
-          end
-        }
-        groups.each { |group|
-          case group
-          when Integer
-            resource = catalog.resources.find { |r| r.is_a?(Puppet::Type.type(:group)) && r.should(:gid) == group }
-            if resource
-              autos << resource
+      if self[:ensure] != :absent
+        if @parameters[:gid]&.shouldorig
+          groups_in_catalog = catalog.resources.filter { |r| r.is_a?(Puppet::Type.type(:group)) }
+          autos += @parameters[:gid].shouldorig.filter_map do |group|
+            if (group.is_a?(String) && group.match?(/^\d+$/)) || group.is_a?(Integer)
+              gid = Integer(group)
+              groups_in_catalog.find { |r| r.should(:gid) == gid }
+            else
+              group
             end
-          else
-            autos << group
           end
-        }
+        end
+
+        if @parameters[:groups]&.should
+          autos += @parameters[:groups].should.split(",")
```

Comment:
can multiple groups be passed as comma separated string?

Reply:
Good question, was just refactoring. Now I question why it was there. I did notice in the other example that `shouldorig` was an array while `should` wasn't. So perhaps that's the problem?

Reply:
Internally puppet stores the desired values as an array (both `should` and `shouldorig`) for example https://github.com/puppetlabs/puppet/blob/7ca202bd2faf876e1904a8a2e891c10c2cee6ffc/lib/puppet/property.rb#L552C1-L552C1

However, the `should` reader will either return a single value (default case) or an array depending on whether the property is single or multi-valued: https://github.com/puppetlabs/puppet/blob/7ca202bd2faf876e1904a8a2e891c10c2cee6ffc/lib/puppet/property.rb#L531-L535

The behavior is determined based on the `array_matching` value specified when the property is defined using `newproperty`. For example, `members` of the group is multi-valued:

https://github.com/puppetlabs/puppet/blob/7ca202bd2faf876e1904a8a2e891c10c2cee6ffc/lib/puppet/type/group.rb#L84